File: integration-test.ts
/**
 * Command integration tests for safe-find and safe-fd
 */

import { assert, assertEquals } from "@std/assert";

async function runCommand(
  cmd: string,
  args: string[],
): Promise<{ success: boolean; output: string; stderr: string }> {
  try {
    const command = new Deno.Command(cmd, {
      args: args,
      stdout: "piped",
      stderr: "piped",
    });

    const result = await command.output();

    return {
      success: result.code === 0,
      output: new TextDecoder().decode(result.stdout),
      stderr: new TextDecoder().decode(result.stderr),
    };
  } catch (error) {
    return {
      success: false,
      output: "",
      stderr: error instanceof Error ? error.message : String(error),
    };
  }
}

Deno.test("safe-find integration - executes with safe options", async () => {
  const result = await runCommand("deno", [
    "run",
    "--allow-run",
    "safe-find.ts",
    ".",
    "-name",
    "*.ts",
    "-type",
    "f",
  ]);

  assert(
    result.success,
    "safe-find should execute successfully with safe options",
  );
  assert(result.output.length > 0, "safe-find should find TypeScript files");
});

Deno.test("safe-find integration - blocks -exec option", async () => {
  const result = await runCommand("deno", [
    "run",
    "--allow-run",
    "safe-find.ts",
    ".",
    "-exec",
    "echo",
    "test",
    ";",
  ]);

  assertEquals(
    result.success,
    false,
    "safe-find should fail when -exec is used",
  );
  assert(
    result.stderr.includes("Dangerous option") &&
      result.stderr.includes("-exec"),
    "Error message should mention dangerous -exec option",
  );
});

Deno.test("safe-find integration - blocks -delete option", async () => {
  const result = await runCommand("deno", [
    "run",
    "--allow-run",
    "safe-find.ts",
    ".",
    "-name",
    "*.tmp",
    "-delete",
  ]);

  assertEquals(
    result.success,
    false,
    "safe-find should fail when -delete is used",
  );
  assert(
    result.stderr.includes("Dangerous option") &&
      result.stderr.includes("-delete"),
    "Error message should mention dangerous -delete option",
  );
});

Deno.test("safe-fd integration - executes with safe options", async () => {
  const result = await runCommand("deno", [
    "run",
    "--allow-run",
    "safe-fd.ts",
    "--glob",
    "*.ts",
  ]);

  assert(
    result.success,
    "safe-fd should execute successfully with safe options",
  );
  assert(result.output.length > 0, "safe-fd should find TypeScript files");
});

Deno.test("safe-fd integration - blocks -x option", async () => {
  const result = await runCommand("deno", [
    "run",
    "--allow-run",
    "safe-fd.ts",
    "--glob",
    "*.ts",
    "-x",
    "echo",
  ]);

  assertEquals(result.success, false, "safe-fd should fail when -x is used");
  assert(
    result.stderr.includes("Dangerous option") && result.stderr.includes("-x"),
    "Error message should mention dangerous -x option",
  );
});

Deno.test("safe-fd integration - blocks --list-details option", async () => {
  const result = await runCommand("deno", [
    "run",
    "--allow-run",
    "safe-fd.ts",
    "--glob",
    "*.ts",
    "--list-details",
  ]);

  assertEquals(
    result.success,
    false,
    "safe-fd should fail when --list-details is used",
  );
  assert(
    result.stderr.includes("Dangerous option") &&
      result.stderr.includes("--list-details"),
    "Error message should mention dangerous --list-details option",
  );
});

File: tests/integration_test.rs
use std::process::Command;

/// Helper function to run commands and capture output
fn run_command(bin_name: &str, args: &[&str]) -> (bool, String, String) {
    let mut cmd = Command::new("cargo");
    cmd.arg("run").arg("--bin").arg(bin_name);

    if !args.is_empty() {
        cmd.arg("--").args(args);
    }

    let output = cmd.output().expect("Failed to execute command");

    let success = output.status.success();
    let stdout = String::from_utf8_lossy(&output.stdout).to_string();
    let stderr = String::from_utf8_lossy(&output.stderr).to_string();

    (success, stdout, stderr)
}

#[test]
fn test_safe_find_with_safe_options() {
    let (success, stdout, _stderr) =
        run_command("safe-find", &[".", "-name", "*.rs", "-type", "f"]);

    assert!(
        success,
        "safe-find should execute successfully with safe options"
    );
    assert!(!stdout.is_empty(), "safe-find should find Rust files");
}

#[test]
fn test_safe_find_blocks_exec_option() {
    let (success, _stdout, stderr) = run_command("safe-find", &[".", "-exec", "echo", "test", ";"]);

    assert!(!success, "safe-find should fail when -exec is used");
    assert!(
        stderr.contains("Dangerous option") && stderr.contains("-exec"),
        "Error message should mention dangerous -exec option"
    );
}

#[test]
fn test_safe_find_blocks_delete_option() {
    let (success, _stdout, stderr) = run_command("safe-find", &[".", "-name", "*.tmp", "-delete"]);

    assert!(!success, "safe-find should fail when -delete is used");
    assert!(
        stderr.contains("Dangerous option") && stderr.contains("-delete"),
        "Error message should mention dangerous -delete option"
    );
}

#[test]
fn test_safe_find_blocks_execdir_option() {
    let (success, _stdout, stderr) =
        run_command("safe-find", &[".", "-execdir", "echo", "test", ";"]);

    assert!(!success, "safe-find should fail when -execdir is used");
    assert!(
        stderr.contains("Dangerous option") && stderr.contains("-execdir"),
        "Error message should mention dangerous -execdir option"
    );
}

#[test]
fn test_safe_fd_with_safe_options() {
    let (success, stdout, _stderr) = run_command("safe-fd", &["--glob", "*.rs"]);

    assert!(
        success,
        "safe-fd should execute successfully with safe options"
    );
    assert!(!stdout.is_empty(), "safe-fd should find Rust files");
}

#[test]
fn test_safe_fd_blocks_exec_option() {
    let (success, _stdout, stderr) = run_command("safe-fd", &["--glob", "*.rs", "-x", "echo"]);

    assert!(!success, "safe-fd should fail when -x is used");
    assert!(
        stderr.contains("Dangerous option") && stderr.contains("-x"),
        "Error message should mention dangerous -x option"
    );
}

#[test]
fn test_safe_fd_blocks_exec_long_option() {
    let (success, _stdout, stderr) = run_command("safe-fd", &["--glob", "*.rs", "--exec", "echo"]);

    assert!(!success, "safe-fd should fail when --exec is used");
    assert!(
        stderr.contains("Dangerous option") && stderr.contains("--exec"),
        "Error message should mention dangerous --exec option"
    );
}

#[test]
fn test_safe_fd_blocks_list_details_option() {
    let (success, _stdout, stderr) = run_command("safe-fd", &["--glob", "*.rs", "--list-details"]);

    assert!(!success, "safe-fd should fail when --list-details is used");
    assert!(
        stderr.contains("Dangerous option") && stderr.contains("--list-details"),
        "Error message should mention dangerous --list-details option"
    );
}

#[test]
fn test_safe_find_help_shows_wrapper_info() {
    let (success, stdout, _stderr) = run_command("safe-find", &["--help"]);

    assert!(success, "safe-find --help should execute successfully");
    assert!(
        stdout.contains("safe-find") && stdout.contains("secure wrapper"),
        "Help output should show safe-find wrapper information"
    );
}

#[test]
fn test_safe_fd_version_shows_wrapper_info() {
    let (success, stdout, _stderr) = run_command("safe-fd", &["--version"]);

    assert!(success, "safe-fd --version should execute successfully");
    assert!(
        stdout.contains("safe-fd") && stdout.contains("secure wrapper"),
        "Version output should show safe-fd wrapper information"
    );
}
